tools/extract-documentation.py

```diff
@@ -0,0 +1,227 @@
+                out += generate_emit_JSON(emit)
+                out += ","
+    if out[-1] != "[":
+        out = out[:-1]
+    out += "],"
+    out += '"location":{}'.format(generate_location_JSON(on.node.location))
+    out += "}"
+    return out
+
+
+def generate_reactor_JSON(reactor):
+    """
+    Generates JSON for a NUClear reactor statement
+
+    Args:
+        on (analyse.struct.Reactor): A structure with information about a reactor statement
+
+    Returns:
+        str: The JSON representation of the reactor statement information
+    """
+    out = "{"
+    out += '"name":"{}",'.format(reactor.get_name())
+    out += '"on":['
+    for method in reactor.methods:
+        for on in method.ons:
+            out += generate_on_JSON(on) + ","
+    if out[-1] != "[":
+        out = out[:-1]
+    out += "],"
+    out += '"location":{}'.format(generate_location_JSON(reactor.node.location))
+    out += "}"
+    return out
+
+
+def generate_module_JSON(module, name):
+    """
+    Generates JSON for a NUClearRoles Module
+
+    Args:
+        on (analyse.struct.Module): A structure with information about a Module
+
+    Returns:
+        str: The JSON representation of the Module information
+    """
+    out = "{"
+    out += '"name":"{}",'.format(name)
+    out += '"reactors":['
+    for reactor in module.reactors:
+        out += generate_reactor_JSON(reactor)
+        out += ","
+    if out[-1] != "[":
+        out = out[:-1]
+    out += "]}"
+    return out
+
+
+def parse(module, files, outdir):
+    """
+    Generates JSON for a module and writes it to the file _reactors.json
+
+    This also uses the lock that is passed into the process at initProcess to stop concurrent writes to the file
+
+    Args:
+        module (str): Module directory
+        files (list[str]): A list of files in the module directory to parse
+        outdir (str): The path to the directory to write the file _reactors.json to
+    """
+    print("Working on module", module)
+    out = generate_module_JSON(analyse.parse_module(files, module), module)
+    lock.acquire()
+    try:
+        with open(os.path.join(outdir, "_reactors" + ".json"), "a") as to_write:
+            to_write.write(out + ",")
+    finally:
+        lock.release()
+
+
+def initProcess(l):
+    """
+    Stores the lock globally in the process
+
+    Args:
+        l (Lock): A lock for writing the output file
+    """
+    global lock
+    lock = l
+
+
+@run_on_docker
+def register(command):
+    command.help = "Generates documentation in JSON format for NUClear reactors and NUClear on and emit statements"
+
+    command.add_argument(
+        "--outdir",
+        default="doc/extract-documentation/",
+        help="The output directory. Default doc/extract-documentation/",
+    )
+
+    command.add_argument(
+        "--indir", default="module", help="The root of the directories that you want to scan. Default: module."
+    )
+
+    command.add_argument(
+        "-m",
```

Comment:
is it possible to make it default to `-m$(nproc)` or equivalent, if `-m` is specified without a number arg? this might end up more effort than its worth, but I would guess that if you're explicitly choosing `-m`, you're going to want to run it as `-m$(nproc)` basically every time anyway.